### DISCUSSION TITLE
Best workflow to deal with multiple layers and events

### DISCUSSION THREAD

**Discussion by iamvdo** [openlayers/openlayers#15386] [General]
Hello everyone,

I'm struggling to find the best approach to manage multiple layers and multiple events on them.
The best method for me would be to add needed events on layers, but the API only allows from map. 

Let's take a simple example: two vector layers with points as circle, each layer should be clickable, but only one at the same time. The possibilities are:

- first option: add select interaction on click for these specific layers. Simplest method, but the 2 layers have to be on same interaction if I want to select only one at the same time, and it's not always feasible (layers can be added in the future) as there is no way to add a new layer to interaction afterwards
- second option: add click event on map, use `getFeaturesAtPixel`, get the first feature (on top). OK, done.

But, let's say I add another vector layer with polygons, on top. Polygons are styled with strokes, so I can see through them (no fill). I don't want them to be clickable. The second option from before returns those polygons.

Sometimes, we're not dealing with features at all (eg, using `layer.getData(evt.pixel)` on a WMS layer), so it becomes really tricky when vector features overlap.

And so on. 
Here is a just a simple case, but in application with multiple layers, unpredictable order, it can become really tough.

How's your solutions? Am I missing something?

Thanks in advance for your responses.
Vincent

**Comment by ahocevar** [comment]
Are you aware of `VectorLayer#getFeatures(pixel)`?

**Comment by iamvdo** [comment]
Not really :) Seems a good addition, but how to deal with first point (eg, select only one feature from multiple layers at a time?) Or say differently, how to be sure to select only the top most visible feature on map (from a random layer)?
